Closes #1550 Initial Intelligence Bank integration.

trackleft opened the pull request:
## Description
Adds the ability to download images from the DAM and manage as either remote media or local media.

## Related issues
Closes #1550 

Spike: https://github.com/az-digital/az_quickstart/issues/1313
EPIC: https://github.com/az-digital/az_quickstart/issues/1551
MVP Issue: https://github.com/az-digital/az_quickstart/issues/1550
MVP PR https://github.com/az-digital/az_quickstart/pull/1592
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1686
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1689
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1688



## How to test
Enable az_dam module.
Go to /admin/content/media
Click on  [+ Add DAM assets]
In the url field click the custom url checkbox
In the url field add assets.arizona.edu
click sign in 
Click sign in with net id
Close once you are signed in, go back to the Add DAM assets page and click continue, or sign in and continue
Now you should be able to find an image and Download it.



## Types of changes
New DAM integration feature
### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [x] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

trackleft commented on modules/custom/az_dam/config/quickstart/ib_dam.settings.yml:
```diff
@@ -0,0 +1,6 @@
+  enforced:
+    module:
+      - az_dam
+debug: true
```
false

trackleft replied:
```suggestion
debug: false
```

joshuasosa commented:
I tested this locally on DDEV and encountered a white screen of death when clicking to Download an image. I also tried it on the ProboCI site and got the same issue. Looking at reports gave some error info.

> Unable to save local file:
> This value should be of the correct primitive type.

> TypeError: Argument 1 passed to Drupal\ib_dam\AssetFormatter\AssetFormatterManager::create() must implement interface Drupal\ib_dam\Asset\AssetInterface, bool given, called in /src/az-quickstart-scaffolding/web/modules/contrib/intelligencebank/modules/ib_dam_media/src/Plugin/EntityBrowser/Widget/IbDamEbWidget.php on line 328 in Drupal\ib_dam\AssetFormatter\AssetFormatterManager::create() (line 29 of /src/az-quickstart-scaffolding/web/modules/contrib/intelligencebank/src/AssetFormatter/AssetFormatterManager.php)

After encountering that error, I tried clicking 'back' on my browser and it took me back to the screen to login with my credentials, which is annoying. It'd be better if it stayed connected.

Also, a related issue I'd have with the usability of this integration is that assets in the "University Photo Library" aren't organized well and generally don't provide useful titles to be able to search on. I saw some Old Main photos on the first page and tried searching for Old Main, but got no results. Having a repository of 6k images with poor searchability makes that part of the repository almost useless for users wanting to find a relevant image. A title of "Now_41" or "UA Campus (25)" gives me no context to know what I'm looking at or what I'd even search for to find that sort of image. Pagination is very slow, which makes just browsing through the images a tedious user experience.

Even after resolving errors, until the user experience is improved for actually finding relevant assets at least through better contextual titles, I'd suggest not enabling this feature request by default.

trackleft commented:
I plan on at least attempting creating a patch against the intelligence bank project on drupal.org that solves the login shenanigans.

joeparsons commented on modules/custom/az_dam/az_dam.info.yml:
```diff
@@ -0,0 +1,8 @@
+name: 'Quickstart DAM'
+type: module
+description:  'The connector for the centralized digital asset management system (IntelligenceBank DAM).'
```
```suggestion
description: 'The connector for the centralized digital asset management system (IntelligenceBank DAM).'
```

trackleft commented:
Move to a follow-up for allowing cdn usage.
```
/**
 * Implements hook_form_alter().
 *
 * Add custom submit handler to form.
 * Allow using remote media.
 * See
 * @link https://git.drupalcode.org/project/intelligencebank/-/blob/8.x-2.x/modules/ib_dam_media/src/Plugin/EntityBrowser/Widget/IbDamEbWidget.php#L249 @endlink.
 */
function az_dam_form_entity_browser_az_dam_form_alter(&$form, FormStateInterface $form_state, $form_id) {
  $form['widget']['ib_dam_app_el']['#allow_embed'] = TRUE;
  $widget_context = $form_state->get(['entity_browser', 'widget_context']);
  $widget_context['ib_dam_media']['allow_embed'] = TRUE;
  $form_state->set(['entity_browser', 'widget_context'], $widget_context);
  $form['#submit'][] = 'az_dam_form_submit';
}
```

trackleft commented:
Updated name of module from "Quickstart DAM" to "Quickstart Arizona Digital Asset Library" and added to "experimental modules" group

Removed remote asset hosting functionality (for now)

Instructions for enabling
Go to `/admin/modules`
Enable Quickstart Digital Asset Library
<img width="1651" alt="image" src="https://user-images.githubusercontent.com/1023167/175115410-2b6a063e-a4b5-493d-937c-37702791507a.png">
<img width="1294" alt="image" src="https://user-images.githubusercontent.com/1023167/175115749-01fe5661-8e0a-49a2-bca9-274bcf8c52d1.png">

Still need to work out permissions, but that is fairly trivial.

Go to `/admin/content/media`
<img width="1298" alt="image" src="https://user-images.githubusercontent.com/1023167/175115982-97868c58-c258-4308-91e9-b5c4bced1c06.png">

Click on "+ Add Arizona Digital Assets"
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/1023167/175116177-1a75df35-c88d-4a0f-8811-d1f76448f975.png">

Select an asset
<img width="1140" alt="image" src="https://user-images.githubusercontent.com/1023167/175116381-8f10fcd7-f38c-4473-a79b-77620e8d9ed2.png">

camikazegreen commented on modules/custom/az_digital_asset_library/config/quickstart/ib_dam_media.settings.yml:
```diff
@@ -0,0 +1,20 @@
+  file:
+    source_type: file
+    media_type: az_document
+  embed:
+    source_type: embed
+    media_type: ib_dam_embed
```
I can add my approval after this change is committed.

joeparsons commented:
Would be good to merge main into this one when possible.

camikazegreen commented on modules/custom/az_digital_asset_library/config/quickstart/ib_dam.settings.yml:
```diff
@@ -0,0 +1,6 @@
+    module:
+      - az_digital_asset_library
+debug: false
+staging: true
```
```suggestion
staging: false
```
According to Garrett, the staging environment should no longer be required.
